##About Boston Dataset !
* **Origin** - The origin of the boston housing data is Natural.
* **Usage** - This dataset may be used for Assessment.
* **Number of Cases** - The dataset contains a total of 506 cases.
* **Order** - The order of the cases is mysterious.
* **Variables** - There are 14 attributes in each case of the dataset.
  They are:
   1.  CRIM - per capita crime rate by town
   2. ZN - proportion of residential land zoned for lots over 25,000 sq.ft
   3. INDUS - proportion of non-retail business acres per town.
   4. CHAS - Charles River dummy variable (1 if tract bounds river; 0 otherwise
   5. NX - nitric oxides concentration (parts per 10 million)
   6. RM - average number of rooms per dwelling
   7. AGE - proportion of owner-occupied units built prior to 1940
   8. DIS - weighted distances to five Boston employment centres
   9. RAD - index of accessibility to radial highways
   10. TAX - full-value property-tax rate per $\$$10,000
   11. PTRATIO - pupil-teacher ratio by town
   12. B - $1000(Bk - 0.63)^2$ where Bk is the proportion of blacks by town
   13. LSTAT - % lower status of the population
   14. MEDV - Median value of owner-occupied homes in $1000's

In [1]:
import pandas as pd
df=pd.read_csv("https://github.com/YBIFoundation/Dataset/raw/main/Boston.csv")
df

CRIM    ZN  INDUS  CHAS     NX     RM   AGE     DIS  RAD    TAX  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296.0   
1    0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242.0   
2    0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242.0   
3    0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222.0   
4    0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222.0   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...    ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273.0   
502  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273.0   
503  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273.0   
504  0.10959   0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273.0   
505  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273.0   

     PTRATIO       B  LSTAT  MEDV  
0       15.3  396.90   4.98  24.0  
1       17.8  396.90   9.14  21.6  
2       17.8  392.83   4.03  34.7  
3       18.7  394.63   2.94  33.4  
4       18.7  396.90   5.33  36.2  
..       ...     ...    ...   ...  
501     21.0  391.99   9.67  22.4  
502     21.0  396.90   9.08  20.6  
503     21.0  396.90   5.64  23.9  
504     21.0  393.45   6.48  22.0  
505     21.0  396.90   7.88  11.9  

[506 rows x 14 columns]

In [2]:
df.isna().sum()

CRIM       0
ZN         0
INDUS      0
CHAS       0
NX         0
RM         0
AGE        0
DIS        0
RAD        0
TAX        0
PTRATIO    0
B          0
LSTAT      0
MEDV       0
dtype: int64

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    int64  
 4   NX       506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    int64  
 9   TAX      506 non-null    float64
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
 13  MEDV     506 non-null    float64
dtypes: float64(12), int64(2)
memory usage: 55.5 KB


In [4]:
df.describe()

CRIM          ZN       INDUS        CHAS          NX          RM  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean     3.613524   11.363636   11.136779    0.069170    0.554695    6.284634   
std      8.601545   23.322453    6.860353    0.253994    0.115878    0.702617   
min      0.006320    0.000000    0.460000    0.000000    0.385000    3.561000   
25%      0.082045    0.000000    5.190000    0.000000    0.449000    5.885500   
50%      0.256510    0.000000    9.690000    0.000000    0.538000    6.208500   
75%      3.677083   12.500000   18.100000    0.000000    0.624000    6.623500   
max     88.976200  100.000000   27.740000    1.000000    0.871000    8.780000   

              AGE         DIS         RAD         TAX     PTRATIO           B  \
count  506.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.574901    3.795043    9.549407  408.237154   18.455534  356.674032   
std     28.148861    2.105710    8.707259  168.537116    2.164946   91.294864   
min      2.900000    1.129600    1.000000  187.000000   12.600000    0.320000   
25%     45.025000    2.100175    4.000000  279.000000   17.400000  375.377500   
50%     77.500000    3.207450    5.000000  330.000000   19.050000  391.440000   
75%     94.075000    5.188425   24.000000  666.000000   20.200000  396.225000   
max    100.000000   12.126500   24.000000  711.000000   22.000000  396.900000   

            LSTAT        MEDV  
count  506.000000  506.000000  
mean    12.653063   22.532806  
std      7.141062    9.197104  
min      1.730000    5.000000  
25%      6.950000   17.025000  
50%     11.360000   21.200000  
75%     16.955000   25.000000  
max     37.970000   50.000000

In [5]:
df.columns

Index(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX',
       'PTRATIO', 'B', 'LSTAT', 'MEDV'],
      dtype='object')

In [6]:
y=df['MEDV']
x=df[['CRIM', 'ZN', 'INDUS', 'CHAS', 'NX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX',
       'PTRATIO', 'B', 'LSTAT']]

In [7]:
import statsmodels.api as sm
x=sm.add_constant(x)

In [8]:
print(sm.OLS(y,x).fit().summary())

                            OLS Regression Results                            
Dep. Variable:                   MEDV   R-squared:                       0.741
Model:                            OLS   Adj. R-squared:                  0.734
Method:                 Least Squares   F-statistic:                     108.1
Date:                Sat, 06 Apr 2024   Prob (F-statistic):          6.72e-135
Time:                        12:57:46   Log-Likelihood:                -1498.8
No. Observations:                 506   AIC:                             3026.
Df Residuals:                     492   BIC:                             3085.
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         36.4595      5.103      7.144      0.0

In [9]:
new_y=df['MEDV']
new_x=df[['CRIM', 'ZN',  'CHAS', 'NX', 'RM',  'DIS', 'RAD', 'TAX',
       'PTRATIO', 'B', 'LSTAT']]

In [32]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(new_x,new_y,random_state=2529)

In [33]:
from sklearn.linear_model import LinearRegression
reg=LinearRegression()

In [34]:
reg.fit(x_train,y_train)

LinearRegression()

In [35]:
y_pred=reg.predict(x_test)
y_pred

array([31.63646872, 22.01325513, 21.06975275, 39.64162546, 20.24543673,
       22.42818768, 18.38598822, 14.95131377, 22.71973043, 21.46241326,
       18.41466575, 27.99607483, 29.52450528,  6.24927313, 10.86571044,
       26.26226127, 21.85582436, 25.17450415,  3.68830111, 36.20861217,
       24.11154504, 23.14325698, 14.33138891, 20.81320773, 24.55420167,
       17.15563391, 18.8230072 , 21.20928643, 28.54510826, 20.85827005,
        8.75274401, 17.21192596, 22.27835392, 22.32728725, 38.97133802,
       26.18125199, 42.154251  , 19.64485015, 33.95788269, 14.10720977,
       13.89763537, 23.00871629, 12.01763894,  8.9543087 , 21.83779929,
       25.33881852, 18.33306874, 17.00885109, 14.56941266, 15.28488661,
       33.52403279, 33.21069155, 15.62940597, 23.96054536, 27.47917475,
       19.5608196 , 44.78236812, 20.77344673, 20.19079075, 27.67097262,
       34.42751384, 12.75973272, 23.56466504, 31.68733292, 28.66885882,
       32.20031813, 13.69782113, 35.30531679, 19.53604279, 19.32

In [36]:
from sklearn.metrics import mean_absolute_error,mean_absolute_percentage_error,mean_squared_error,r2_score
mean_absolute_error(y_test,y_pred)

3.2474163606364463

In [37]:
mean_absolute_percentage_error(y_test,y_pred)

0.1701301749325574

In [38]:
mean_squared_error(y_test,y_pred)

22.778291970320918

In [39]:
r2_score(y_test,y_pred)

0.6514600203265164

**Conclusion:-** We need some more features for prediction as the given features can only explain the model only 65.15%. Hence, given features are not enough.